src/algorithms/linear_stretching.jl

```diff
@@ -39,44 +43,132 @@ channel are stretched to the specified range. The modified Y channel is then
+    end
+end
+function LinearStretching(rangemap::Pair{Tuple{T1, T2}, Tuple{T3, T4}}) where {T1, T2, T3, T4}
+    LinearStretching(rangemap.first..., rangemap.second...)
+end
+function LinearStretching(rangemap::Pair{Nothing, Tuple{T3, T4}}) where {T3, T4}
+    LinearStretching(nothing, nothing, rangemap.second...)
+end
+function LinearStretching(src_range::Tuple{T1, T2}) where {T1, T2}
+    LinearStretching(src_minval=src_range[1], src_maxval=src_range[2])
 end
 
 function (f::LinearStretching)(out::GenericGrayImage, img::GenericGrayImage)
-    src_minval = minfinite(img)
-    src_maxval = maxfinite(img)
+    img_min, img_max = minfinite(img), maxfinite(img)
+    src_minval = isnothing(f.src_minval) ? img_min : f.src_minval
+    src_maxval = isnothing(f.src_maxval) ? img_max : f.src_maxval
+    dest_minval = f.dest_minval
+    dest_maxval = f.dest_maxval
     T = eltype(out)
-    out .= img
-    map!(out,out) do val
+
+    # the kernel operation `r * x - o` is equivalent to `(x-A) * ((b-a)/(B-A)) + a`
+    # precalculate these and make inner loop contains only multiplication and addition
+    # to get better performance
+    r = convert(floattype(T), (dest_maxval - dest_minval) / (src_maxval - src_minval))
+    o = convert(floattype(T), (src_minval*dest_maxval - src_maxval*dest_minval) / (src_maxval - src_minval))
+
+    if 1 ≈ r && 0 ≈ o
+        # when image intensity is already adjusted, there's no need to do it again
+        # it's a trivial but common case in practice
+        out === img || (out .= img)
+        return out
+    end
+
+    # In most cases, we don't need to clamp the output
+    # this is only used when user specifies custom `(src_minval, src_maxval)`
+    out_minval = r * img_min - o
+    out_maxval = r * img_max - o
+    do_clamp = (out_minval < dest_minval) || (out_maxval > dest_maxval)
+
+    # early convert type to hit faster clamp version
+    #  -- this might not be the root reason but it gives performance difference locally
+    dest_minval = convert(typeof(out_minval), dest_minval)
+    dest_maxval = convert(typeof(out_maxval), dest_maxval)
+
+    # tweak the performance of FixedPoint by fusing operations into one broadcast
+    # for Float32 the fallback implementation is faster
+    if eltype(T) <: FixedPoint
+        # ?: is faster than if-else
+        @. out = do_clamp ? clamp(r * img - o, dest_minval, dest_maxval) : r * img - o
+        return out
+    end
+
+    # fallback implementation
+    @inbounds @simd for p in eachindex(img)
+        val = img[p]
         if isnan(val)
-            return val
+            out[p] = val
         else
-            newval = linear_stretch(val, src_minval, src_maxval, f.minval, f.maxval)
-            return  T <: Integer ? round(Int, newval ) : newval
+            newval = r * val - o
+            do_clamp && (newval = clamp(newval, dest_minval, dest_maxval))
+            out[p] = T <: Integer ? round(Int, newval) : newval
```

Comment:
For `Gray{Float32}` input, 47b127b avoids large memory allocation in this line by making `newval` to be `Gray{Float32}` instead of `Float32` ( 🤔 `Gray(0.1) - 0 === 0.1` instead of `Gray(0.1)`)

@kimikage do you happen to know what's the issue here?

Reply:
I think I've figured it out via `@code_warntype`; that's caused by unstable type inference.